library/src/main/java/eu/kanade/tachiyomi/source/ResolvableSource.kt

```diff
@@ -0,0 +1,42 @@
+package eu.kanade.tachiyomi.source
+
+import eu.kanade.tachiyomi.source.model.SChapter
+import eu.kanade.tachiyomi.source.model.SManga
+import okhttp3.HttpUrl
+
+/**
+ * Represents a source that can resolve a URL into a specific type of [Resource].
+ */
+interface ResolvableSource {
+
+    /**
+     * Checks whether the given [url] can be resolved by this source.
+     *
+     * @param url The [HttpUrl] to test.
+     * @return `true` if this source can resolve the URL, `false` otherwise.
+     */
+    fun canResolve(url: HttpUrl): Boolean
```

Comment:
Just make `getResource` return `Resource?` and remove this method.

Source might need to make network requests to know if they can resolve the thing. Most cases when they instantly return `false`, the `suspend` won't cause overhead.

Reply:
this is mainly for early filtering on app side. should just be a regular regex check on source side

Reply:
1. `getResouce` also needs the regex/URL path check to extract path segments. Your API forces extensions to repeat this logic.
2. The app just calls the first then the second. What's your use case that only calls the first one?

Reply:
Returning `null` is equivalent to "can't resolve". No results found should be conveyed by throwing an exception.

Reply:
What you're proposing is that we show user Source A, B and C. When Source C returns `null` we remove it from the list. Later Source A and B resolves to Manga A. I don't think that's a good UX. There should be a way for the app to filter out sources that 100% will never resolve the URL